consensus/src/dag/dag_driver.rs

```diff
@@ -96,37 +100,62 @@ impl DagDriver {
                 .read()
                 .get_strong_links_for_round(current_round, &driver.epoch_state.verifier)
                 .unwrap_or(vec![]);
-            driver.enter_new_round(current_round + 1, strong_links);
+            block_on(driver.enter_new_round(current_round + 1, strong_links));
         }
         driver
     }
 
-    pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
-        let mut dag_writer = self.dag.write();
-        let round = node.metadata().round();
+    pub async fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        let maybe_strong_links = {
+            let mut dag_writer = self.dag.write();
+            let round = node.metadata().round();
 
-        if !dag_writer.all_exists(node.parents_metadata()) {
-            if let Err(err) = self.fetch_requester.request_for_certified_node(node) {
-                error!("request to fetch failed: {}", err);
+            if !dag_writer.all_exists(node.parents_metadata()) {
+                if let Err(err) = self.fetch_requester.request_for_certified_node(node) {
+                    error!("request to fetch failed: {}", err);
+                }
+                bail!(DagDriverError::MissingParents);
             }
-            bail!(DagDriverError::MissingParents);
-        }
 
-        dag_writer.add_node(node)?;
-        if self.current_round == round {
-            let maybe_strong_links = dag_writer
-                .get_strong_links_for_round(self.current_round, &self.epoch_state.verifier);
-            drop(dag_writer);
-            if let Some(strong_links) = maybe_strong_links {
-                self.enter_new_round(self.current_round + 1, strong_links);
+            dag_writer.add_node(node)?;
+            if self.current_round == round {
+                dag_writer
+                    .get_strong_links_for_round(self.current_round, &self.epoch_state.verifier)
+            } else {
+                None
             }
+        };
+
+        if let Some(strong_links) = maybe_strong_links {
+            self.enter_new_round(self.current_round + 1, strong_links)
+                .await;
         }
         Ok(())
     }
 
-    pub fn enter_new_round(&mut self, new_round: Round, strong_links: Vec<NodeCertificate>) {
+    pub async fn enter_new_round(&mut self, new_round: Round, strong_links: Vec<NodeCertificate>) {
+        debug!("entering new round {}", new_round);
         // TODO: support pulling payload
-        let payload = Payload::empty(false);
+        let payload = match self
+            .payload_client
+            .pull_payload(
+                Duration::from_secs(1),
+                100,
+                1000,
+                PayloadFilter::Empty,
```

Comment:
We need a filter. 
I think we forgot about this functionality, we need to maintain a data structure with all the batches currently in local DAG. (Or we can traverse the dag each time before pulling.)

Reply:
yes, need to add a TODO. @zekun000  is taking care of it.